File: backend/app/schemas/prompt.py
from datetime import datetime
from pydantic import BaseModel, ConfigDict


class AuthorOut(BaseModel):
    """프롬프트 응답에 포함될 작성자 정보"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    username: str


class PromptBase(BaseModel):
    type: str
    keyword: str
    ai_model: str | None = None
    content: str
    description: str | None = None
    is_hide: bool = False


class PromptCreate(PromptBase):
    """POST /api/prompts 요청 바디"""
    thumbnail_url: str | None = None


class PromptRead(PromptBase):
    """프롬프트 응답 (목록/상세 공용)"""
    model_config = ConfigDict(from_attributes=True)

    id: str
    ai_model: str | None = None
    rank: int
    thumbnail_url: str | None = None
    views: int
    like_count: int = 0  # 추가된 필드
    created_at: datetime
    author: AuthorOut
    is_liked: bool = False


# 하위호환용 별칭
PromptOut = PromptRead


class LikeToggleResponse(BaseModel):
    """좋아요 토글 시 프론트엔드로 전달할 lightweight 응답"""
    is_liked: bool
    like_count: int


class PromptListResponse(BaseModel):
    prompts: list[PromptRead]
    page: int = 1
    page_size: int = 12
    total: int = 0
    total_pages: int = 1


class KeywordItem(BaseModel):
    keyword: str
    prompt_count: int = 0


class KeywordListResponse(BaseModel):
    keywords: list[str]

File: frontend/types/prompt.ts
export type PromptType = "image" | "text";

export type PromptAuthor = {
    id: number;
    username: string;
};

/** 백엔드 API에서 넘어오는 Raw 응답 규격 (snake_case) */
export type PromptApi = {
    id: number;
    type: string;
    keyword: string;
    ai_model: string | null;
    rank: number;
    content: string;
    description: string | null;
    thumbnail_url: string | null;
    views: number;
    like_count?: number;
    is_liked?: boolean;
    is_hide?: boolean;
    created_at: string;
    author: PromptAuthor | string;
};

/** 프론트엔드 UI 컴포넌트 전체에서 사용할 단일 표준 모델 (camelCase) */
export type Prompt = {
    id: number;
    type: PromptType;
    keyword: string;
    aiModel?: string;
    rank: number;
    content: string;
    description?: string;
    thumbnailUrl?: string;
    author: PromptAuthor;
    views: number;
    likeCount: number;
    isLiked: boolean;
    createdAt: string;
    isHide?: boolean;
};

export type PromptListResponse = {
    prompts: PromptApi[];
    page: number;
    page_size: number;
    total: number;
    total_pages: number;
};

export type PromptPage = {
    prompts: Prompt[];
    page: number;
    pageSize: number;
    total: number;
    totalPages: number;
};

export type PromptQuery = {
    keyword?: string;
    type?: PromptType;
    sort?: "rank" | "recent";
    page?: number;
    pageSize?: number;
};

export type PromptCreateForm = {
    keyword: string;
    aiModel: string;
    content: string;
    description: string;
    thumbnailFile: File | null;
    captcha_token: string;
};
